contracts/rewards/RewardsManager.sol

```diff
@@ -0,0 +1,178 @@
+    }
+
+    /// @notice Update the given pool's SUSHI allocation point and `IRewarder` contract. Can only be called by the owner.
+    /// @param _pool The address of the pool. See `poolInfo`.
+    /// @param _allocPoint New AP of the pool.
+    /// @param _rewarder Address of the rewarder delegate.
+    /// @param overwrite True if _rewarder should be `set`. Otherwise `_rewarder` is ignored.
+    function set(
+        address _pool,
+        uint256 _allocPoint,
+        IRewarder _rewarder,
+        bool overwrite
+    ) public onlyOwner {
+        PoolInfo memory info = poolInfo[_pool];
+        if (info.lastRewardBlock == 0 && info.allocPoint == 0) {
+            totalAllocPoint = totalAllocPoint + _allocPoint;
+            poolInfo[_pool].lastRewardBlock = uint64(block.number);
+        } else {
+            totalAllocPoint = (totalAllocPoint - poolInfo[_pool].allocPoint) + _allocPoint;
+        }
+
+        poolInfo[_pool].allocPoint = uint64(_allocPoint);
+
+        if (overwrite) {
+            rewarder[_pool] = _rewarder;
+        }
+
+        emit LogSetPool(_pool, _allocPoint, overwrite ? _rewarder : rewarder[_pool], overwrite);
+    }
+
+    /// @notice Update reward variables for all pools. Be careful of gas spending!
+    /// @param pools Pool addresses of all to be updated. Make sure to update all active pools.
+    function massUpdatePools(IPool[] calldata pools) external {
+        uint256 len = pools.length;
+        for (uint256 i = 0; i < len; ++i) {
+            updatePool(pools[i]);
+        }
+    }
+
+    /// @notice Harvest rewards for a specific account for a given pool.
+    /// @param pool The address of the pool. See `poolInfo`.
+    /// @param account The account to claim for.
+    function claimRewardsFor(IPool pool, address account) external {
+        PoolInfo memory info = updatePool(pool);
+
+        int256 debt = rewardDebt[address(pool)][account];
+        uint256 amount = IERC20(address(pool)).balanceOf(account);
+
+        int256 accumulatedSushi = int256((amount * info.accSushiPerShare) / ACC_SUSHI_PRECISION);
+        uint256 _pendingSushi = uint256(accumulatedSushi - debt);
+
+        // Effects
+        rewardDebt[address(pool)][account] = accumulatedSushi;
+
+        // Interactions
+        if (_pendingSushi != 0) {
+            SUSHI.transfer(account, _pendingSushi);
+        }
+
+        IRewarder _rewarder = rewarder[address(pool)];
+        if (address(_rewarder) != address(0)) {
+            RewardAmount memory reward = unclaimedRewards[address(pool)][account];
+            try _rewarder.onSushiReward(address(pool), account, account, _pendingSushi + reward.pendingSushi, amount + reward.amount) {
+                unclaimedRewards[address(pool)][account] = RewardAmount({pendingSushi: 0, amount: 0});
+            } catch {
+                unclaimedRewards[address(pool)][account] = RewardAmount({pendingSushi: reward.pendingSushi + _pendingSushi, amount: reward.amount + amount});
```

Comment:
Imagine if a user has 10 LP tokens. If they call it once while the rewarder fails, on the next call the rewarder is called with a new LP amount of 20, even though the user still only has 10 LP tokens. If the event is e.g. a transfer it is also problematic because the rewards continue to run for two accounts.

Reply:
Yeah we need to think about how best to do this. Maybe subtract the rewards.amount from pendingRewards if its greater than 0?